src/System.Net.Sockets/tests/FunctionalTests/SocketOptionNameTest.cs

```diff
@@ -284,6 +284,88 @@ public void ReuseAddress_Windows(bool? exclusiveAddressUse, bool? firstSocketReu
             ReuseAddress(exclusiveAddressUse, firstSocketReuseAddress, secondSocketReuseAddress, expectFailure);
         }
 
+        [Fact]
+        [PlatformSpecific(TestPlatforms.AnyUnix)] // Windows defaults are different
+        public void ExclusiveAddress_Default_Unix()
+        {
+            using (Socket a = new Socket(AddressFamily.InterNetwork, SocketType.Dgram, ProtocolType.Udp))
+            {
+                Assert.Equal(1, (int)a.GetSocketOption(SocketOptionLevel.Socket, SocketOptionName.ExclusiveAddressUse));
+                Assert.Equal(true, a.ExclusiveAddressUse);
+                Assert.Equal(0, (int)a.GetSocketOption(SocketOptionLevel.Socket, SocketOptionName.ReuseAddress));
+            }
+        }
+
+        [Theory]
+        [InlineData(1)]
+        [InlineData(0)]
+        [PlatformSpecific(TestPlatforms.AnyUnix)] // Unix does not have separate options for ExclusiveAddressUse and ReuseAddress.
+        public void SettingExclusiveAddress_SetsReuseAddress(int value)
+        {
+            using (Socket a = new Socket(AddressFamily.InterNetwork, SocketType.Dgram, ProtocolType.Udp))
+            {
+                a.SetSocketOption(SocketOptionLevel.Socket, SocketOptionName.ExclusiveAddressUse, value);
+
+                Assert.Equal(value, (int)a.GetSocketOption(SocketOptionLevel.Socket, SocketOptionName.ExclusiveAddressUse));
+                Assert.Equal(value == 1 ? 0 : 1, (int)a.GetSocketOption(SocketOptionLevel.Socket, SocketOptionName.ReuseAddress));
+            }
+
+            // SettingReuseAddress_SetsExclusiveAddress
+            using (Socket a = new Socket(AddressFamily.InterNetwork, SocketType.Dgram, ProtocolType.Udp))
+            {
+                a.SetSocketOption(SocketOptionLevel.Socket, SocketOptionName.ReuseAddress, value);
+
+                Assert.Equal(value, (int)a.GetSocketOption(SocketOptionLevel.Socket, SocketOptionName.ReuseAddress));
+                Assert.Equal(value == 1 ? 0 : 1, (int)a.GetSocketOption(SocketOptionLevel.Socket, SocketOptionName.ExclusiveAddressUse));
+            }
+        }
+
+        [Fact]
+        public void BindDuringTcpWait_Succeeds()
+        {
+            int port = 0;
+            using (Socket a = new Socket(AddressFamily.InterNetwork, SocketType.Stream, ProtocolType.Tcp))
+            {
+                a.Bind(new IPEndPoint(IPAddress.Loopback, 0));
+                port = (a.LocalEndPoint as IPEndPoint).Port;
+                a.Listen(10);
+
+                // Connect a client
+                using (Socket client = new Socket(AddressFamily.InterNetwork, SocketType.Stream, ProtocolType.Tcp))
+                {
+                    client.Connect(new IPEndPoint(IPAddress.Loopback, port));
+                    // accept socket and close it with zero linger time.
+                    a.Accept().Close(0);
+                }
+            }
+
+            // Bind a socket to the same address we just used.
```

Comment:
Do we run any risk here that another test running concurrently might end up getting assigned this port in the short window between the above and here?  Seems like it could result in random CI failures now and then?

Reply:
That's possible. I don't think there's a way to deal with that. Perhaps the allocation strategy in the dynamic port range makes it very unlikely this will happen. We can see if it causes issues in CI?

Reply:
We do this for [some of our Kestrel tests](https://github.com/aspnet/KestrelHttpServer/blob/64322ae2c5e8749b1976b505f82f01f45d2b0f93/test/Kestrel.FunctionalTests/AddressRegistrationTests.cs#L819-L831). We run this regularly on various versions of Windows, Linux and macOS without that being a source of flakiness.

Reply:
> We run this regularly on various versions of Windows, Linux and macOS without that being a source of flakiness.

Do you run lots of tests in parallel?  I seem to remember this actually being an issue for us in the past, but I could be misremembering.

Reply:
That's a good point. All the tests that use the GetNextPort() method I linked to are in the same test class, so xunit does not parallelize them. We're only just now looking into running multiple test projects in parallel.